tasks/run_agent.rb

```diff
@@ -0,0 +1,160 @@
+#!/opt/puppetlabs/puppet/bin/ruby
+
+require 'open3'
+require 'etc'
+require 'json'
+require 'yaml'
+require_relative '../../ruby_task_helper/files/task_helper.rb'
+
+class RunAgentTask < TaskHelper
+  def task(puppet_bin: '/opt/puppetlabs/bin/puppet', wait_time: 300, _noop: false, **_kwargs)
+    @puppet_bin = puppet_bin
+    @wait_time = wait_time
+    @noop = _noop
+
+    check_running_as_root
+    check_agent_not_disabled
+    wait_for_puppet_lockfile
+    run_puppet
+  end
+
+  def wait_for_puppet_lockfile
+    waited = 0
+    while agent_locked? && waited < @wait_time
+      sleep 1
+      waited += 1
+    end
+
+    if agent_locked?
+      raise TaskHelper::Error.new(
+        'Lockfile still exists after waiting',
+        'theforeman-puppet/lockfile_timeout_expired'
+      )
+    end
+
+    waited
+  end
+
+  def check_agent_not_disabled
+    if agent_disabled?
+      raise TaskHelper::Error.new(
+        'Agent is disabled on this node',
+        'theforeman-puppet/agent_disabled',
+        disabled_reason
+      )
+    end
+  end
+
+  def agent_locked?
+    File.exist?(agent_catalog_run_lockfile)
+  end
+
+  def agent_disabled?
+    File.exist?(agent_disabled_lockfile)
+  end
+
+  def disabled_reason
+    file = File.open agent_disabled_lockfile
+    data = JSON.load file
+    file.close
+    data
+  end
+
+  def agent_disabled_lockfile
+    return @agent_disabled_lockfile unless @agent_disabled_lockfile.nil?
+
+    @agent_disabled_lockfile = agent_config('agent_disabled_lockfile')
+    @agent_disabled_lockfile
+  end
+
+  def agent_catalog_run_lockfile
+    return @lockfile unless @lockfile.nil?
+
+    @lockfile = agent_config('agent_catalog_run_lockfile')
+    @lockfile
+  end
+
+  def lastrunfile
+    return @lastrunfile unless @lastrunfile.nil?
+
+    @lastrunfile = agent_config('lastrunfile')
+    @lastrunfile
+  end
+
+  def agent_config(setting)
+    cmd = [@puppet_bin, 'config', 'print', setting]
+    stdout, stderr, status = Open3.capture3(*cmd)
+
+    unless status.exitstatus.zero?
+      raise TaskHelper::Error.new(
+        "Couldn't determine #{setting} configuration",
+        'theforeman-puppet/config_unknown',
+        stderr: stderr
+      )
+    end
+    stdout.strip
+  end
+
+  def check_running_as_root
+    unless Process.euid.zero?
+      raise TaskHelper::Error.new(
+        'Puppet agent needs to run as root',
+        'theforeman-puppet/bad_euid',
+        euid: Process.euid
+      )
+    end
+  end
+
+  def run_puppet
+    cmd = [@puppet_bin, 'agent', '-t', '--detailed-exitcodes']
+    cmd << if @noop
+             '--noop'
+           else
+             '--no-noop'
```

Comment:
This is to override the value in the config file.  I should probably make this clearer in whatever docs I write or in the task metadata.